### ISSUE TITLE
Make use of Kotlin Contracts

### ISSUE THREAD

**Issue by MartinHaeusler** [robfletcher/strikt#200]
For example, [`callsInPlace`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.contracts/-contract-builder/calls-in-place.html) would be very useful (I assume that this is actually the case in strikt?). Maybe even [`implies`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.contracts/-simple-effect/implies.html) for `.isNotNull()` checks, if it is possible.

For example, I would much rather write:

`expectThat(someValue).isNotNull()`

... than:

`assertNotNull(someValue)`

But `assertNotNull` has the big advantage that it uses a contract, telling the compiler that the varialbe is no longer nullable afterwards. I am aware of the `.and { ... }` method in strikt, but sometimes I need to call something, assert something, call something else and assert again. I don't know if it is even possible to make use of an `implies` contract here (due to the chained method calls) but it could be worth looking into it.

**Comment by MartinHaeusler** [comment]
Seems like I spoke too soon: I just did some quick testing with contracts, and they are not officially out of the experimental phase. The consequence is that all code directly or transitively calling a method with a custom contract needs to be annotated as experimental. I guess it's just too soon to think about actually using custom contracts.

I think it's still worthwhile to keep this in mind for later though.

**Comment by robfletcher** [comment]
Definitely. I had a brief look when contracts were first announced, and it's something I'd like to take advantage of in the long term if possible. I'm not sure that contracts as they stand can be used to imply something in a method (e.g. `isNotNull`) that applies to the parameter of a different method (`expectThat`). I probably need to do a little more research. I _think_ `callsInPlace` might be useful for `expectCatching`, etc.
